Add runtime migration creation and application

thromel opened the pull request:
## Summary

Implements #37342: Allow creating and applying migrations at runtime without recompiling.

This adds support for creating and applying migrations at runtime using Roslyn compilation, enabling scenarios like .NET Aspire and containerized applications where recompilation isn't possible.

### CLI Usage

```bash
# Standard update (existing behavior)
dotnet ef database update [migration]

# Create and apply a new migration in one step
dotnet ef database update MigrationName --add [--output-dir <DIR>] [--namespace <NS>] [--json]
```

The `-o/--output-dir`, `-n/--namespace`, and `--json` options require `--add` to be specified.

### PowerShell Usage

```powershell
# Standard update (existing behavior)
Update-Database [-Migration <migration>]

# Create and apply a new migration in one step
Update-Database -Migration MigrationName -Add [-OutputDir <DIR>] [-Namespace <NS>]
```

### Architecture

| Component | Purpose |
|-----------|---------|
| `IMigrationCompiler` / `CSharpMigrationCompiler` | Internal: Roslyn-based compilation of scaffolded migrations |
| `IMigrationsAssembly.AddMigrations(Assembly)` | Registers dynamically compiled migrations |
| `MigrationsOperations.AddAndApplyMigration()` | Orchestrates scaffold → compile → register → apply workflow |

### Design Decisions

- **Extends existing services**: Uses `IMigrationsScaffolder` for scaffolding and `IMigrator` for applying, adding only the new `IMigrationCompiler` service
- **AddMigrations(Assembly)**: Extended `IMigrationsAssembly` interface to accept additional assemblies containing runtime-compiled migrations
- **Always persists to disk**: Like `AddMigration`, files are always saved to enable source control and future recompilation
- **No pending changes behavior**: If no model changes are detected, applies any existing pending migrations without creating a new one
- **Internal compiler API**: `IMigrationCompiler` and `CSharpMigrationCompiler` are in the `.Internal` namespace as they require design work for public API
- **Error handling with cleanup**: If compilation or migration application fails, saved migration files are cleaned up to prevent orphans
- **Thread safety**: `MigrationsAssembly` uses locking to protect against race conditions when adding migrations concurrently

### Workflow

```
User runs: dotnet ef database update InitialCreate --add
    │
    ▼
MigrationsOperations.AddAndApplyMigration()
    │
    ├─► Check for pending model changes
    │       └─► If none: apply existing migrations, return
    │
    ├─► IMigrationsScaffolder.ScaffoldMigration() - Generate code
    │
    ├─► try {
    │       ├─► IMigrationsScaffolder.Save() - Write files to disk
    │       ├─► IMigrationCompiler.CompileMigration() - Roslyn compile
    │       ├─► IMigrationsAssembly.AddMigrations() - Register migration
    │       └─► IMigrator.Migrate() - Apply to database
    │   } catch {
    │       └─► Clean up saved files on failure
    │   }
    │
    └─► Return migration files
```

### Robustness Features

1. **Exception handling with cleanup**: `AddAndApplyMigration` wraps the save-compile-register-apply chain in try-catch, deleting saved files on failure to prevent orphans
2. **Context disposal on validation failure**: `PrepareForMigration` ensures the DbContext is disposed if validation or service building fails
3. **Thread-safe migration registration**: `MigrationsAssembly` uses locking to protect shared state (migrations dictionary, model snapshot, additional assemblies list)

### Limitations

- Requires dynamic code generation (incompatible with NativeAOT) - marked with `[RequiresDynamicCode]`
- C# only (no VB.NET/F# support)

## Test plan

- [x] Unit tests for `CSharpMigrationCompiler`
- [x] Unit tests for `MigrationsOperations.AddAndApplyMigration`
- [x] Integration tests in `RuntimeMigrationTestBase` (SQLite and SQL Server implementations)
- [x] Tests for validation (empty name, invalid characters)
- [x] Tests for RemoveMigration with dynamically created migrations
- [x] All existing EFCore.Design.Tests pass
- [x] All existing EFCore.Relational.Tests pass

Fixes #37342

thromel commented:
## Note on SQL Server Integration Tests

The `RuntimeMigrationSqlServerTest` tests are marked with `[SqlServerCondition(SqlServerCondition.IsNotAzureSql | SqlServerCondition.IsNotCI)]` and are skipped in CI. This follows the same pattern used by `MigrationsInfrastructureSqlServerTest`.

**Why these tests are skipped in CI:**
- They require creating fresh databases dynamically for each test to properly test the migration flow from scratch
- The Helix CI environment has SQL Server available but with limited permissions configured for shared/pre-configured databases
- Tests that use `SqlServerTestStore.CreateInitializedAsync` with dynamic database names don't work in the CI SQL Server setup

**Test coverage is still maintained:**
- Core runtime migration logic is covered by unit tests in `EFCore.Design.Tests` (which run in CI)
- The SQL Server integration tests run locally for developers with SQL Server configured
- SQLite integration tests in `EFCore.Sqlite.FunctionalTests` also validate the end-to-end flow

This is consistent with how other complex migration infrastructure tests handle CI limitations.

AndriySvyryd commented on test/EFCore.SqlServer.FunctionalTests/Migrations/RuntimeMigrationSqlServerTest.cs:
```diff
@@ -0,0 +1,344 @@
+
+    public async Task InitializeAsync()
+    {
+        _testStore = await SqlServerTestStore.CreateInitializedAsync(
```
These tests don't need to always run from scratch, an empty database will be as good to test the added functionality. Use a non-dynamic database name and don't delete the database files.
Don't call `EnsureDeleted`, `CreateInitializedAsync` already cleans the database

AndriySvyryd commented on src/ef/IOperationExecutor.cs:
```diff
@@ -15,6 +15,7 @@ internal interface IOperationExecutor : IDisposable
     void DropDatabase(string? contextType);
     IDictionary GetContextInfo(string? name);
     void UpdateDatabase(string? migration, string? connectionString, string? contextType);
+    IDictionary CreateAndApplyMigration(string name, string? connectionString, string? contextType, string? outputDir, string? @namespace);
```
```suggestion
    IDictionary AddAndApplyMigration(string name, string? outputDir, string? contextType, string? @namespace, string? connectionString);
```

AndriySvyryd commented on src/EFCore.Design/Design/Internal/MigrationsOperations.cs:
```diff
@@ -272,6 +272,62 @@ public virtual void HasPendingModelChanges(string? contextType)
+        {
+            throw new OperationException(
+                DesignStrings.ConflictingContextAndMigrationName(name));
+        }
```
Extract common validation

AndriySvyryd commented on src/EFCore.Design/Migrations/Design/IDynamicMigrationsAssembly.cs:
```diff
@@ -0,0 +1,38 @@
+    /// <summary>
+    ///     Gets a value indicating whether any dynamic migrations have been registered.
+    /// </summary>
+    bool HasDynamicMigrations { get; }
```
Remove these as they don't appear to be used and merge this interface into `IMigrationsAssembly`

AndriySvyryd commented on src/ef/Commands/DatabaseUpdateCommand.cs:
```diff
@@ -1,16 +1,45 @@
+            {
+                throw new CommandException(Resources.NamespaceRequiresAdd);
+            }
+        }
```
It's better to do this in the `Validate()` override

AndriySvyryd commented on src/EFCore.Relational/Extensions/RelationalDatabaseFacadeExtensions.cs:
```diff
@@ -197,6 +197,80 @@ public static Task MigrateAsync(
+    [RequiresDynamicCode(
+        "Runtime migration compilation requires dynamic code generation. "
+        + "This operation is not supported with NativeAOT.")]
+    public static RuntimeMigrationResult CreateAndApplyMigration(
```
Remove these methods. This functionality is only available through the `EFC.Design` assembly, it shouldn't be exposed here.

AndriySvyryd commented on src/EFCore.Design/Migrations/Design/RuntimeMigrationService.cs:
```diff
@@ -0,0 +1,489 @@
+///         See <see href="https://aka.ms/efcore-docs-migrations">Database migrations</see> for more information and examples.
+///     </para>
+/// </remarks>
+public class RuntimeMigrationService : IRuntimeMigrationService
```
Don't create a new service for this, distribute the functionality among the existing services like `IMigrationsScaffolder` and `IMigrator`,

AndriySvyryd commented on src/ef/Commands/DatabaseUpdateCommand.cs:
```diff
@@ -1,16 +1,49 @@
+            {
+                throw new CommandException(Resources.NamespaceRequiresAdd);
+            }
+        }
```
~Also throw if `_migration` was specified~

AndriySvyryd replied:
Actually, no. Instead make `_add` a switch and use `_migration` value as the name for the new migration

AndriySvyryd commented on src/ef/Commands/DatabaseUpdateCommand.cs:
```diff
@@ -1,16 +1,49 @@
+        if (_add!.HasValue())
+        {
+            // Create and apply a new migration in one step
+            executor.AddAndApplyMigration(
```
I see that you populate the result for the created migration, but don't actually use it, see https://github.com/dotnet/efcore/blob/2cc3da841692bb6348c7a589b9af2ec9e36d67d9/src/ef/Commands/MigrationsAddCommand.cs#L39 for how it can be output

AndriySvyryd commented on src/ef/Commands/DatabaseUpdateCommand.cs:
```diff
@@ -1,16 +1,49 @@
 // ReSharper disable once ArrangeTypeModifiers
 internal partial class DatabaseUpdateCommand
 {
+    protected override void Validate()
```
Move this to the `DatabaseUpdateCommand.Configure.cs‎` file so that it's also included in dotnet-ef

AndriySvyryd commented on src/ef/Commands/DatabaseUpdateCommand.Configure.cs:
```diff
@@ -18,6 +21,9 @@ public override void Configure(CommandLineApplication command)
         _connection = command.Option("--connection <CONNECTION>", Resources.DbContextConnectionDescription);
+        _add = command.Option("--add <NAME>", Resources.DatabaseUpdateAddDescription);
+        _outputDir = command.Option("-o|--output-dir <PATH>", Resources.MigrationsOutputDirDescription);
+        _namespace = command.Option("-n|--namespace <NAMESPACE>", Resources.MigrationsNamespaceDescription);
```
We try to keep the PMC tooling at feature parity with the `dotnet ef` command, so please add matching parameters to `Update-Database` in https://github.com/dotnet/efcore/blob/2cc3da841692bb6348c7a589b9af2ec9e36d67d9/src/EFCore.Tools/tools/EntityFrameworkCore.psm1

AndriySvyryd commented on src/EFCore.Design/Migrations/Design/IMigrationCompiler.cs:
```diff
@@ -0,0 +1,36 @@
+///         See <see href="https://aka.ms/efcore-docs-migrations">Database migrations</see> for more information and examples.
+///     </para>
+/// </remarks>
+public interface IMigrationCompiler
```
This would require some design work the get the API right. For now, move it and `CSharpMigrationCompiler` to the `.Internal` namespace and change the XML documentation accordingly.

AndriySvyryd commented on src/EFCore.Relational/Migrations/Internal/MigrationsAssembly.cs:
```diff
@@ -53,34 +55,49 @@ public virtual IReadOnlyDictionary<string, TypeInfo> Migrations
         get
         {
-            IReadOnlyDictionary<string, TypeInfo> Create()
+            lock (_lock)
```
Instead of using `lock` call `NonCapturingLazyInitializer.EnsureInitialized`

AndriySvyryd commented on src/EFCore.Design/Design/Internal/MigrationsOperations.cs:
```diff
@@ -272,6 +254,157 @@ public virtual void HasPendingModelChanges(string? contextType)
+    {
+        var (resolvedOutputDir, subNamespace, context, services) = PrepareForMigration(name, outputDir, contextType);
+
+        using (context)
```
Convert to a using declaration to reduce nesting

AndriySvyryd commented on src/EFCore.Design/Design/Internal/MigrationsOperations.cs:
```diff
@@ -272,6 +257,114 @@ public virtual void HasPendingModelChanges(string? contextType)
+        if (string.IsNullOrWhiteSpace(name))
+        {
+            throw new OperationException(DesignStrings.MigrationNameRequired);
+        }
```
Remove this here and from `AddMigration` as `PrepareForMigration` is already checking this.

AndriySvyryd commented on test/EFCore.Relational.Specification.Tests/RuntimeMigrationTestBase.cs:
```diff
@@ -0,0 +1,931 @@
+            => "RuntimeMigration";
+
+        protected override bool UsePooling
+            => false;
```
Why did you have to specify `false` here?

thromel commented:
Addressed the remaining review feedback:

1. **`context.Database.OpenConnection()` / `CloseConnection()`** - Updated all direct `connection.Open()`/`Close()` calls to use the EF Core methods

2. **`Fixture.ReseedAsync()`** - Implemented `IAsyncLifetime` on the test class and call `Fixture.ReseedAsync()` in `InitializeAsync`. Moved the cleanup logic to the fixture's `CleanAsync` override.

3. **Lock removal from `AddMigrations`** - Done in previous commit

Commits: e9b3d543c7, a5e5331a65

AndriySvyryd commented on src/EFCore.Design/Design/Internal/MigrationsOperations.cs:
```diff
@@ -272,6 +257,114 @@ public virtual void HasPendingModelChanges(string? contextType)
+        if (string.IsNullOrWhiteSpace(name))
+        {
+            throw new OperationException(DesignStrings.MigrationNameRequired);
+        }
```
Ok, then remove this check since it's already checked before this call

thromel commented on src/EFCore.Relational/Migrations/Internal/MigrationsAssembly.cs:
```diff
@@ -155,4 +190,18 @@ public virtual Migration CreateMigration(TypeInfo migrationClass, string activeP
+        _additionalAssemblies.Add(additionalMigrationsAssembly);
+        _migrations = null;
+        _modelSnapshot = null;
+        _modelSnapshotInitialized = false;
```
This was discussed with @AndriySvyryd. The conclusion was that the race condition is very unlikely since `AddMigrations` is only called during design-time operations, not concurrently with `Migrations` property access. The lock was intentionally removed per reviewer feedback. See comment: https://github.com/dotnet/efcore/pull/37415#issuecomment-3726457860

thromel commented on src/ef/Commands/DatabaseUpdateCommand.Configure.cs:
```diff
@@ -18,7 +22,36 @@ public override void Configure(CommandLineApplication command)
+            if (_namespace!.HasValue())
+            {
+                throw new CommandException(Resources.NamespaceRequiresAdd);
+            }
```
Fixed in commit a5b2659e1e - added validation that `--json` requires `--add` to be specified, consistent with `--output-dir` and `--namespace`.

AndriySvyryd commented on test/EFCore.Design.Tests/Design/Internal/MigrationsOperationsTest.cs:
```diff
@@ -43,6 +45,46 @@ public void Can_use_migrations_assembly()
+    [ConditionalFact]
+    public void AddMigration_throws_when_name_is_empty()
+    {
+        var assembly = MockAssembly.Create(typeof(TestContext));
```
Use `AssemblyTestContext` in these tests

AndriySvyryd commented on src/ef/Commands/DatabaseUpdateCommand.cs:
```diff
@@ -10,8 +13,35 @@ protected override int Execute(string[] args)
+
+            if (_json!.HasValue())
+            {
+                ReportJson(files);
```
If `files` is empty or just contains null values don't report anything.

AndriySvyryd commented on src/EFCore.Design/Design/Internal/MigrationsOperations.cs:
```diff
@@ -272,6 +252,111 @@ public virtual void HasPendingModelChanges(string? contextType)
+        catch (Exception ex)
+        {
+            throw new OperationException(
+                DesignStrings.AddAndApplyMigrationFailed(name, ex.Message), ex);
```
You don't need to wrap this exception, so remove the try/catch

AndriySvyryd commented on src/EFCore.Design/Design/Internal/MigrationsOperations.cs:
```diff
@@ -272,6 +252,111 @@ public virtual void HasPendingModelChanges(string? contextType)
+        MigrationFiles? files = null;
+        try
+        {
+            files = scaffolder.Save(_projectDir, migration, resolvedOutputDir, dryRun: false);
```
Move this, as well as `scaffolder` declaration to a new method that can also be called by `AddMigration`, make sure to preserve the code comment about #18950.

`outputDir` and `subNamespace` calculations can be moved from `PrepareForMigration` to this new shared method.

AndriySvyryd commented on src/EFCore.Design/Migrations/Design/CSharpMigrationCompiler.cs:
```diff
@@ -0,0 +1,183 @@
+    /// <returns>The list of metadata references.</returns>
+    protected virtual IReadOnlyList<MetadataReference> GetMetadataReferences(
+        Type contextType,
+        IEnumerable<Assembly>? additionalReferences)
```
It seems that `additionalReferences` is always `null`, so it can be removed.

AndriySvyryd commented on src/EFCore.Design/Migrations/Design/CSharpMigrationCompiler.cs:
```diff
@@ -0,0 +1,183 @@
+        var allReferences = new List<MetadataReference>(baseReferences);
+
+        // Add the context's assembly (in case it wasn't loaded when cache was built)
+        AddAssemblyReference(allReferences, contextType.Assembly);
```
I can't see a situation where the context assembly wouldn't be in the cache, so you can remove this call and this whole method can then be inlined

AndriySvyryd commented on src/EFCore.Design/Migrations/Design/IMigrationCompiler.cs:
```diff
@@ -0,0 +1,39 @@
+    /// <param name="contextType">The type of the <see cref="DbContext" /> for which the migration was created.</param>
+    /// <param name="references">Additional assembly references to include in compilation, if any.</param>
+    /// <returns>An <see cref="Assembly" /> containing the compiled migration and model snapshot.</returns>
+    /// <exception cref="InvalidOperationException">Thrown when compilation fails.</exception>
```
```suggestion
/// <summary>
///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
///     the same compatibility standards as public APIs. It may be changed or removed without notice in
///     any release. You should only use it directly in your code with extreme caution and knowing that
///     doing so can result in application failures when updating to a new Entity Framework Core release.
/// </summary>
```

AndriySvyryd commented on src/EFCore.Relational/Migrations/Internal/MigrationsAssembly.cs:
```diff
@@ -112,12 +121,38 @@ orderby id
+    {
+        get
+        {
+            lock (_lock)
```
Use `NonCapturingLazyInitializer.EnsureInitialized` here instead of lock

AndriySvyryd commented on src/EFCore.Relational/Migrations/Internal/MigrationsAssembly.cs:
```diff
@@ -121,30 +121,24 @@ orderby id
-                return _modelSnapshot;
-            }
-        }
+        return snapshot ?? NoSnapshot;
```
Just return `snapshot` if it's `null`. `NullModelSnapshot` and `NoSnapshot` can be removed
